EXPERIMENT: JSON:API backend (#40832)
discourse/discourse: plugins/discourse-data-explorer/docs/rfc-draft.md

# RFC: A versioned JSON:API for Discourse

> This is the RFC phase of the [REST API overhaul project](https://dev.discourse.org/t/rest-api-overhaul-project-overview/187625): what the new API looks like, how it's used and how it's written. Everything here has been prototyped in the Data Explorer plugin ([PR #40832](https://github.com/discourse/discourse/pull/40832)), the exploration itself is logged in [Modernizing how we write APIs in Discourse](https://dev.discourse.org/t/modernizing-how-we-write-apis-in-discourse-a-json-api-experiment/186394), and the full design with the reasoning behind each decision is in reference docs in the PR. I link them from the relevant sections below.

Reviewer: This is quite a dense paragraph to start with. Maybe nice to split it up a bit - could do something like:
```suggestion
This is the RFC for our [REST API overhaul](https://dev.discourse.org/t/rest-api-overhaul-project-overview/187625).

Prior work:
- Prototype in data-explorer plugin: ([PR #40832](https://github.com/discourse/discourse/pull/40832)), including full design & reasoning in markdown files

- Exploration logs: [Modernizing how we write APIs in Discourse](https://dev.discourse.org/t/modernizing-how-we-write-apis-in-discourse-a-json-api-experiment/186394)
```

If you have suggestions/corrections, please post below, and I’ll handle integrating any changes.

> 

Reviewer: ### Tl;dr:

- New api under `/api/...`
- Using JSON:API formats for requests and response
- AMS serializers are replaced with a Resource object
- Versioned by date. Controllers only implement latest version. AR-style migrations are used to support older requests & responses
- All requests, responses and version-changes are self-documenting
- Plugins can add namespaced relations to core resources, but not modify their attributes

## Background

Our APIs are hand-rolled: AMS serializers, ad-hoc pagination, and a different response shape per endpoint, with related data side-loaded differently everywhere (see [this post](https://dev.discourse.org/t/modernizing-how-we-write-apis-in-discourse-a-json-api-experiment/186394/1)). Avoiding breaking changes is a best effort, and we fail regularly. And the documentation is written by hand, separately from the code, so it doesn’t always stay up to date (see [documentation update](https://dev.discourse.org/t/modernizing-how-we-write-apis-in-discourse-a-json-api-experiment/186394/19)).

> So instead of defining our own conventions, we're adopting JSON:API. It's a frozen spec, and it already standardizes what we do (more or less) by hand today: compound documents (`included` plus linkage), `include`, sparse fieldsets, `filter`, `sort` and `page`. It's also what WarpDrive is normalized around. And as its documents are type-tagged, with standard error pointers and reserved parameter families, the versioning machinery below stays relatively small.

Reviewer: Shall we link JSON:API to https://jsonapi.org/ ?

For some of the 'jargon', I wonder if some footnotes would be useful. Especially for:
- compound documents (I guess it means 'api responses which contain data from multiple models'?)

- `include` is the same as compound documents? Or something else?

- 'sparse fieldsets'? Is that the feature which allows clients to ask for only certain fields?

- 'type-tagged, with standard error pointers and reserved parameter families' - not 100% sure what all of this means at-a-glance?

Also **mega**-tiny nitpick. Maybe say "WarpDrive is **designed** around", instead of "normalized"? A little easier to read, especially for less-technical people.

## Goals

From the [project overview](https://dev.discourse.org/t/rest-api-overhaul-project-overview/187625):

1. Third parties can depend on our REST API without worrying about unexpected breaking changes.
2. Consistent formats for requests and responses, including common patterns like side-loading related data.
3. It’s easy for us (CDCK developers) to create and evolve APIs using these patterns.
4. The API is clearly documented and can be referenced by customer-facing teams as a major feature of Discourse.
5. Designed to avoid fundamental performance issues (e.g. cursor pagination instead of `LIMIT...OFFSET`).
6. Suited to integrate with our frontend app ([WarpDrive migration](https://dev.discourse.org/t/warpdrive-migration-project-overview/187528)).

## For integrators (using the API)

### Authentication

> Nothing changes here, we keep the existing API key credentials (`Api-Key`/`Api-Username`). Authorization doesn't change either and keeps using our guardians: a resource declares the scope a caller can see, attributes can be restricted, and writes go through the usual service framework.

Reviewer: I just thought...  we should probably try to integrate the API Scopes system into our DSL  and docs 👀 

Could go here, or in its own section.

Try to plug into existing system, but with a nicer DSL.

```json
{
  "data": [
    {
      "id": "1",
      "type": "queries",
      "attributes": {
        "name": "Top referred topics",
        "query": "SELECT id, title FROM topics ORDER BY like_count DESC LIMIT 10",
        "ran_at": "2026-07-01T10:00:00.000Z"
      },
      "relationships": {
        "user": { "data": { "id": "1", "type": "users" } },
        "groups": { "data": [{ "id": "1", "type": "groups" }] }
      },
      "meta": { "page": { "cursor": "…" } }
    }
  ],
  "included": [
    { "id": "1", "type": "users", "attributes": { "usernames": ["query_master"] } },
    { "id": "1", "type": "groups", "attributes": { "name": "sql_writers" } }
  ],
  "links": { "prev": null, "next": "…" }
}
```

One request returns the list, its authors and their groups, instead of a request per author. It's also the format WarpDrive consumes natively.

Writes take a JSON:API document too (`data.type` plus `attributes`, relationships as linkage) and follow the spec's methods: `POST` to create, `PATCH` to update.

### Pagination

> Collections use the JSON:API [cursor pagination profile](https://jsonapi.org/profiles/ethanresnick/cursor-pagination): `page[size]`, `page[after]`, `page[before]`, and `links.prev` / `links.next` to follow. There's no offset pagination, as `LIMIT...OFFSET` gets slower when the offset grows. The trade-off for integrators: no page numbers and no jumping to page N, and a total count is requested explicitly (`stats[total]=count`) instead of coming with every page. Explained in the [versioning update](https://dev.discourse.org/t/modernizing-how-we-write-apis-in-discourse-a-json-api-experiment/186394/13).

Reviewer: `stats` specification? Or our own custom thing?

Remove it? Risky for perf anyway?

### Errors

Errors are JSON:API error documents. On listings, unknown filters, sorts, includes or page parameters are rejected with a `400` naming what wasn't recognized, instead of being ignored. Validation failures return `422` with a JSON Pointer per invalid attribute, using the names of the client's pinned version.

### Documentation

The reference documentation is generated from the same declarations that serve the API, so the two can't diverge, and it's versioned: selecting a date shows the API as that pin sees it. There's one document per owner (core, then one per plugin), each with its own changelog. Described in the [documentation update](https://dev.discourse.org/t/modernizing-how-we-write-apis-in-discourse-a-json-api-experiment/186394/19), and you can browse the generated docs for the prototype here:

https://raw.githack.com/discourse/discourse/loic/json-api-experiments/plugins/discourse-data-explorer/openapi-docs.html

## For developers (writing endpoints)

### Resource classes

> A resource class declares the document shape and the query surface in one place, including the scope a caller may see. Attribute types are required: the documentation is generated from them, and it's the same type vocabulary service contracts already use.

Reviewer: ```suggestion
A resource class declares the document shape and the query surface in one place, including the ActiveRecord scope a caller may see. Attribute types are required: the documentation is generated from them, and it's the same type vocabulary service contracts already use.
```

- Exploration logs: [Modernizing how we write APIs in Discourse](https://dev.discourse.org/t/modernizing-how-we-write-apis-in-discourse-a-json-api-experiment/186394)

If you have suggestions/corrections, please post below, and I’ll handle integrating any changes.

## TL;DR

- A new API served under a global `/api/...` namespace (that's the proposal, see the open questions)
- JSON:API formats for requests and responses, so one shape everywhere, with related data side-loaded on request
- AMS serializers are replaced with a resource object: document shape and query surface (filters, sorts, includes, pagination) declared in one place
- Versioned by date, and the version header is mandatory. Controllers only implement the latest version, and dated version changes translate older requests and responses, a bit like AR migrations but for the wire, never for stored data
- Requests, responses and version changes are self-documenting: the reference docs and the changelog are generated from them
- Cursor pagination only: no page numbers, no offsets, no total counts
- Plugins can add namespaced relations and filters to core resources, but not modify their attributes

> - The current API stays in place, the new one is opt-in per request

Reviewer: Some ideas to make the tl;dr even shorter:
```suggestion
- A new API under `/api/...`
- JSON:API formats for requests and responses
- AMS serializers are replaced with a resource object
- Versioned by a mandatory date-based `Api-Version` header
- Controllers only implement the latest version. VersionChange system translates older requests and responses, similar to database migrations
- Docs & changelog are auto-generated from the code
- Cursor pagination only: no page numbers, no offsets, no total counts
- Plugins add relations and filters to core resources, but never modify their attributes
- The current API stays in place (for now)
```
(feel free to cherry-pick/edit/ignore as you like)
